src/control/server/server_utils.go

```diff
@@ -195,20 +171,60 @@ func prepBdevStorage(srv *server, iommuEnabled bool, hpiGetter common.GetHugePag
 		}
 	}
 
-	// Run prepare with reset first to release resources.
-	//
-	// TODO: should be passing root context into prepare request to
-	//       facilitate cancellation.
-	prepReq.Reset_ = true
+	// Perform an automatic prepare based on the values in the config file.
+	prepReq := storage.BdevPrepareRequest{
+		TargetUser:   srv.runningUser,
+		PCIAllowList: strings.Join(srv.cfg.BdevInclude, storage.BdevPciAddrSep),
+		PCIBlockList: strings.Join(srv.cfg.BdevExclude, storage.BdevPciAddrSep),
+		DisableVFIO:  srv.cfg.DisableVFIO,
+		EnableVMD:    srv.cfg.EnableVMD && !srv.cfg.DisableVFIO && iommuEnabled,
+		Reset_:       true, // Run prepare with reset first to release resources.
+	}
+
 	if _, err := srv.ctlSvc.NvmePrepare(prepReq); err != nil {
 		srv.log.Errorf("automatic NVMe prepare reset failed: %s", err)
+	}
+
+	if hasBdevs {
+		// Use default value if bdevs have been configured but nr_hugepages unset in config.
+		if srv.cfg.NrHugepages < 0 {
+			srv.cfg.NrHugepages = 4096
+		}
+
+		prepReq.HugePageCount = srv.cfg.NrHugepages
+
+		// The config value is intended to be per-engine, so we need to allocate hugepages on
+		// each engine's numa node.
+		if srv.cfg.NrHugepages > 0 {
+			var nodes []string
+			for _, ec := range srv.cfg.Engines {
+				nodes = append(nodes, fmt.Sprintf("%d", ec.Storage.NumaNodeIndex))
+			}
+			if len(nodes) != 0 {
+				// Allocate HugePageCount on each numa node in HugeNodes.
+				prepReq.HugeNodes = strings.Join(nodes, ",")
+			}
+		} else {
+			return errors.New("bdevs detected in config but no hugepages requested")
+		}
 	} else {
-		prepReq.Reset_ = false
-		if _, err := srv.ctlSvc.NvmePrepare(prepReq); err != nil {
-			srv.log.Errorf("automatic NVMe prepare failed: %s", err)
+		// If nr_hugepages unset and no bdevs in config, set minimum needed for scanning.
+		if srv.cfg.NrHugepages < 0 {
+			prepReq.HugePageCount = minHugePageCount
```

Comment:
Do we need:

src.cfs.NrHuepages = 0 here?

Reply:
I've added it on line 217 because whenever we don't have bdevs we don't want to allocate hugepages for engines.

Reply:
No, we could use hugepages even without bdevs, see this:
https://daosio.atlassian.net/browse/DAOS-9000

Reply:
ok